projects/smoke_mcp23008/src/main.c

```diff
@@ -0,0 +1,111 @@
+// Smoke test for mcp23008 gpio expander
+// Test every GPIO on the expander
+
+#include "gpio.h"
+#include "i2c.h"
+#include "interrupt.h"
+#include "log.h"
+#include "mcp23008_gpio_expander.h"
+#include "soft_timer.h"
+#include "wait.h"
+
+// wait time between gpio pin toggling
+#define WAIT_TIME_MILLIS 1000
+
+#define MCP23008_I2C_ADDRESS 0x20  // MCP23008 address
+#define I2C_PORT I2C_PORT_2
+
+#define I2C_SCL \
+  { GPIO_PORT_B, 10 }
+#define I2C_SDA \
+  { GPIO_PORT_B, 11 }
+
+void setup(void) {
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,
+    .sda = I2C_SDA,
+    .scl = I2C_SCL,
+  };
+  i2c_init(I2C_PORT, &i2c_settings);
+  mcp23008_gpio_init(I2C_PORT, MCP23008_I2C_ADDRESS);
+}
+
+StatusCode prv_mcp23008_init_all_pins(Mcp23008GpioDirection direction) {
+  Mcp23008GpioSettings gpio_settings = {
+    .direction = direction,
+  };
+  Mcp23008GpioAddress address = {
+    .i2c_address = MCP23008_I2C_ADDRESS,
+  };
+
+  if (direction == MCP23008_GPIO_DIR_OUT) {
+    gpio_settings.state = MCP23008_GPIO_STATE_HIGH;
+  } else {
+    gpio_settings.state = MCP23008_GPIO_STATE_LOW;
+  }
+  for (Mcp23008PinAddress pin = 0; pin < NUM_MCP23008_GPIO_PINS; pin++) {
+    address.pin = pin;
+    status_ok_or_return(mcp23008_gpio_init_pin(&address, &gpio_settings));
+  }
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_mcp23008_check_pin_states(Mcp23008GpioState state) {
+  Mcp23008GpioState in_state;
+  Mcp23008GpioAddress address = { .i2c_address = MCP23008_I2C_ADDRESS };
+  for (Mcp23008PinAddress pin = 0; pin < NUM_MCP23008_GPIO_PINS; pin++) {
+    address.pin = pin;
+    status_ok_or_return(mcp23008_gpio_get_state(&address, &in_state));
+    if (state == in_state) {
+      LOG_DEBUG("State for pin %d set and read correctly\n", pin);
+    } else {
+      LOG_DEBUG("State for pin %d incorrect\n", pin);
+      return STATUS_CODE_INTERNAL_ERROR;
+    }
+  }
+  return STATUS_CODE_OK;
+}
+
+static void prv_soft_timer_callback_output(SoftTimerId timer_id, void *context) {
+  Mcp23008GpioState *state = (Mcp23008GpioState *)context;
+  Mcp23008GpioState get_state;
+  Mcp23008GpioAddress address = { .i2c_address = MCP23008_I2C_ADDRESS };
+
+  for (Mcp23008PinAddress pin = 0; pin < NUM_MCP23008_GPIO_PINS; pin++) {
+    address.pin = pin;
+    mcp23008_gpio_toggle_state(&address);
+    mcp23008_gpio_get_state(&address, &get_state);
+    if (get_state == *state) {
+      LOG_DEBUG("State for pin %d set and read correctly\n", pin);
+    } else {
+      LOG_DEBUG("State for pin %d incorrectly\n", pin);
+    }
+  }
+  *state =
+      (*state == MCP23008_GPIO_STATE_HIGH ? MCP23008_GPIO_STATE_LOW : MCP23008_GPIO_STATE_HIGH);
+
+  LOG_DEBUG("GPIO state = %d\n", *state);
```

Comment:
I think this log should go before you flip the state, since this way you'll toggle gpio, then flip `*state`, then print `*state`, which would then be inaccurate

Reply:
I also thing you should use a string format here to be more clear, something like
```
"GPIO state = %s\n", *state == STATE_HIGH ? "high" : "low"
```
Or something similar